### PR TITLE
feature(form-field) create new component to handle errors and hints

### LINKED ISSUE #294: Brn and Hlm Error Directive for Form Fields
### Which scope/s are relevant/related to the feature request?

Don't know / other

### Information

Error messages for form controls currently need to be handled by user instead of having a directive similar to mat-error to handle showing and hiding the error message to the user and maintaining the style consistent with Shadcn

### Describe any alternatives/workarounds you're currently using

_No response_

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by cjosue15**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our
      guidelines: https://github.com/goetzrobin/spartan/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

- [ ] accordion
- [ ] alert
- [ ] alert-dialog
- [ ] aspect-ratio
- [ ] avatar
- [ ] badge
- [ ] button
- [ ] calendar
- [ ] card
- [ ] checkbox
- [ ] collapsible
- [ ] combobox
- [ ] command
- [ ] context-menu
- [ ] data-table
- [ ] date-picker
- [ ] dialog
- [ ] dropdown-menu
- [x] form-field
- [ ] hover-card
- [ ] icon
- [ ] input
- [ ] label
- [ ] menubar
- [ ] navigation-menu
- [ ] pagination
- [ ] popover
- [ ] progress
- [ ] radio-group
- [ ] scroll-area
- [ ] select
- [ ] separator
- [ ] sheet
- [ ] skeleton
- [ ] slider
- [ ] sonner
- [ ] spinner
- [ ] switch
- [ ] table
- [ ] tabs
- [ ] textarea
- [ ] toast
- [ ] toggle
- [ ] tooltip
- [ ] typography

## What is the current behavior?

None

Closes #294

## What is the new behavior?

Create a new component to handle errors and hints like the material functionality but with the shadcn styles.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
 	"eslint.experimental.useFlatConfig": true,
 	"eslint.format.enable": true,
 	"[typescript]": {
-		"editor.defaultFormatter": "biomejs.biome"
+		"editor.defaultFormatter": "esbenp.prettier-vscode"
```
Comment (goetzrobin) [inline .vscode/settings.json:6]: what's the reason for this? We are switching to biome for formatting. I might have to update the docs on this

Comment (cjosue15) [reply .vscode/settings.json:6]: Sorry, I forgot to delete it

File: libs/ui/core/src/lib/brain/error-options.ts
Comment (goetzrobin) [inline libs/ui/core/src/lib/brain/error-options.ts:1]: Instead of core, I feel like this should be in a separate brain lib for forms. Thoughts?

File: libs/ui/core/src/lib/brain/error-state-tracker.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import type { ErrorStateMatcher } from "./error-options";
+
+export class ErrorStateTracker {
+	/** Whether the tracker is currently in an error state. */
+	errorState = signal(false);
+
+	/** User-defined matcher for the error state. */
+	matcher: ErrorStateMatcher | null = null;
+
+	constructor(
```
Comment (goetzrobin) [inline libs/ui/core/src/lib/brain/error-state-tracker.ts:12]: Let's use the inject function here

File: libs/ui/core/src/lib/brain/error-state-tracker.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { signal } from "@angular/core";
+import type { AbstractControl, FormGroupDirective, NgControl, NgForm } from "@angular/forms";
+import type { ErrorStateMatcher } from "./error-options";
+
+export class ErrorStateTracker {
+	/** Whether the tracker is currently in an error state. */
+	errorState = signal(false);
```
Comment (goetzrobin) [inline libs/ui/core/src/lib/brain/error-state-tracker.ts:7]: I don't love errorState as a name. I feel like I'd expect an object with more information instead of a boolean, maybe something like hasError?

File: libs/ui/core/src/lib/brain/error-options.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { Injectable } from '@angular/core';
+import type { AbstractControl, FormGroupDirective, NgForm } from '@angular/forms';
+
+/** Error state matcher that matches when a control is invalid and dirty. */
+@Injectable()
+export class ShowOnDirtyErrorStateMatcher implements ErrorStateMatcher {
+	isErrorState(control: AbstractControl | null, form: FormGroupDirective | NgForm | null): boolean {
```
Comment (goetzrobin) [inline libs/ui/core/src/lib/brain/error-options.ts:7]: These names should probably be adjusted to fit the variable name

File: libs/ui/form-field/helm/src/lib/tests/form-field-example.ts
Comment (goetzrobin) [inline libs/ui/form-field/helm/src/lib/tests/form-field-example.ts:1]: I think we can remove the tests and directives folder and just have a flat folder structure. Also, let's rename this file to form-field-example.spec.ts, I understand there are not actual tests, but naming it that way definitely ensures it's not bundled for prod builds

File: libs/ui/input/helm/src/lib/hlm-input.directive.ts [hunk 5]
```diff
@@ -49,4 +58,36 @@ export class HlmInputDirective {
...
+
+	errorStateTracker: ErrorStateTracker;
+
+	private defaultErrorStateMatcher = inject(ErrorStateMatcher);
+	private parentForm = inject(NgForm, { optional: true });
+	private parentFormGroup = inject(FormGroupDirective, { optional: true });
+
+	errorState = computed(() => this.errorStateTracker.errorState());
+
+	constructor(@Inject(Injector) private injector: Injector) {
```
Comment (goetzrobin) [inline libs/ui/input/helm/src/lib/hlm-input.directive.ts:72]: Let's not mix the inject function and constructor injector

File: libs/ui/input/helm/src/lib/hlm-input.directive.ts [hunk 6]
```diff
@@ -49,4 +58,36 @@ export class HlmInputDirective {
...
+			() => {
+				if (this.ngControl) {
+					this.error = this.errorStateTracker.errorState();
+				}
+			},
+			{ allowSignalWrites: true },
+		);
+	}
+
+	ngDoCheck() {
```
Comment (goetzrobin) [inline libs/ui/input/helm/src/lib/hlm-input.directive.ts:90]: Why is this necessary? It feels like an ant-ipattern to have to do this in every supported component.

File: libs/ui/select/brain/src/lib/brn-select.component.ts [hunk 7]
```diff
@@ -245,6 +267,10 @@ export class BrnSelectComponent implements ControlValueAccessor, AfterContentIni
 		}
 	}
 
+	ngDoCheck() {
+		this.errorStateTracker.updateErrorState();
```
Comment (goetzrobin) [inline libs/ui/select/brain/src/lib/brn-select.component.ts:271]: Same here

**Review by goetzrobin — CHANGES_REQUESTED** [review]
I added some comments! I really like the idea of this and the overall approach seems fine to me, too. One big question I have: Is there a way to support this with uncontrolled form controls, too? e.g. just add a (host) directive to any form control that's Angular form compatible and automatically get the functionality, instead of having to manually update the form state on every doCheck

**Comment by cjosue15** [comment]
> I added some comments! I really like the idea of this and the overall approach seems fine to me, too. One big question I have: Is there a way to support this with uncontrolled form controls, too? e.g. just add a (host) directive to any form control that's Angular form compatible and automatically get the functionality, instead of having to manually update the form state on every doCheck

When you say uncontrolled form controls, you mean inputs or selects without any of these (FormControl or formcontrolName).
Also I think it is not possible to use this validation as HostDirective because only some components must have the errorState, another point is each component that will be used inside the FormFIeld component should be use the BrnFormFieldControl in the providers, to be identified by the parent(form-field), additionally using the update of the error in the ngDoCheck is the only way, at least that I know of, where I can reevaluate on every change detection when the error is triggered.

File: libs/ui/core/src/lib/brain/error-state-tracker.ts (hunk 2 above)
Comment (cjosue15) [reply libs/ui/core/src/lib/brain/error-state-tracker.ts:12]: I get your point, but this is not a service is just a class.

File: libs/ui/core/src/lib/brain/error-state-tracker.ts (hunk 3 above)
Comment (cjosue15) [reply libs/ui/core/src/lib/brain/error-state-tracker.ts:7]: I can change it, any idea about the content of the object?
You mean something like

```
state = signal<{ hasError: boolean, ... }>({ hasError: false, ... })
```

File: libs/ui/input/helm/src/lib/hlm-input.directive.ts (hunk 5 above)
Comment (cjosue15) [reply libs/ui/input/helm/src/lib/hlm-input.directive.ts:72]: Done!

File: libs/ui/form-field/helm/src/lib/tests/form-field-example.ts
Comment (cjosue15) [reply libs/ui/form-field/helm/src/lib/tests/form-field-example.ts:1]: I removed the folder directives and put the files in the `/lib` folder, but regarding the `test` folder, there are 2 files, one the `form-field.spec.ts` and the `form-field-eample.ts`, the first one is the test itself and the second one is the file containing some mock components for the test file.

I saw the same pattern in other packages as select brain

image

File: libs/ui/core/src/lib/brain/error-options.ts
Comment (cjosue15) [reply libs/ui/core/src/lib/brain/error-options.ts:1]: something like `ui/form/brain/src/lib/error-options.ts` ?

File: libs/ui/core/src/lib/brain/error-options.ts (hunk 4 above)
Comment (cjosue15) [reply libs/ui/core/src/lib/brain/error-options.ts:7]: I think i can change it to `isInValid`? any suggestion?

File: libs/ui/core/src/lib/brain/error-options.ts
Comment (cjosue15) [reply libs/ui/core/src/lib/brain/error-options.ts:1]: any thoughts?

File: libs/ui/input/helm/src/lib/hlm-input.directive.ts (hunk 6 above)
Comment (thatsamsonkid) [reply libs/ui/input/helm/src/lib/hlm-input.directive.ts:92]: I was seeing that material actually does it this way as well. They have an ngDoCheck in their form components and it's usually to call some sort of function to update the error state. One example from their repo https://github.com/angular/components/blob/ca7c090fb948c2f27f13d76e7cb567d0ba870ab9/src/material/select/select.ts#L678 . Otherwise not sure how else we could it.

**Review by thatsamsonkid — COMMENTED** [review]
Just added a few comments for now but overall looks great. Thanks for all the hard work so far, this will be a great addition!

**Comment by cjosue15** [comment]
Hi @thatsamsonkid and @goetzrobin thanks for your time, i think i resolve each doubt, please let me know if you have any question.
One point I would like to understand, is how can I test the generator for this new component, I have created it but I am not sure how to test it.

**Review by elite-benni — APPROVED** [review]

**Review by thatsamsonkid — APPROVED** [review]
LGTM! As far as the generators, I'm not too sure myself. @goetzrobin is there a way to test them before publishing this?

**Comment by cjosue15** [comment]
any thoughts here?, can we merge it? @goetzrobin
